feat: Add support for SOURCE packages

toprinse opened the pull request:
# Motivation

This PR adds support for building packages **from source**, a feature that is currently missing. See issue [#80](https://github.com/Quantco/pixi-pack/issues/80) for background and motivation.

When a local package is found in the pixi.lock of the main package, start a `pixi-build` command to build the local dependencies and add them as injected packages in the `environment.tar`
# Changes

* Introduces a `pixi-build` command to automatically build local packages from the main `pixi.lock`
* Adding integration test to demonstrate the the added code is working as expected.

Closes #80

Commit:
feat: update pixi-pack to be able to pack packages from source

feat: try to add intergation tests

pavelzw committed:
chore: Update shasums after release (#242)

Commit:
feat: add integration tests for local build packages

fix: remove unused fixture

fix: fix typo

fix: fix typo

fix: fix typo

fix: fix test on arm-64, osx and win builds

fix: fix test on arm-64, osx and win builds

fix: add valid checksum for other architectures

fix: fix test using correct checksum

fix: fix typo error

fix: fix typo

delsner commented on src/pack.rs:
```diff
@@ -89,6 +90,100 @@ fn load_lockfile(manifest_path: &Path) -> Result<LockFile> {
+        anyhow::bail!("Failed to build package in {}", manifest_dir.display());
+    }
+    Ok(())
+}
```
`pixi-pack` is an async rust program. Can we please use the async tokio variant of [`Command`](https://docs.rs/tokio/latest/tokio/process/struct.Command.html) here?

toprinse replied:
I fixed using `tokio::process::Command`

delsner commented on src/pack.rs:
```diff
@@ -89,6 +90,100 @@ fn load_lockfile(manifest_path: &Path) -> Result<LockFile> {
+            for v in arr {
+                deps.extend(find_all_local_dependencies(v, manifest_dir, _manifest_path));
+            }
+        }
```
again, why is the recursion necessary? I pushed a minimal example [here](https://github.com/Quantco/pixi-pack/blob/81c120a47e6f6660c179878cc243d1fa29a9c139/examples/local-build/main-pkg/pixi.lock#L29-L34) which shows that the main package's `pixi.lock` already contains all upstream dependencies.

toprinse replied:
You are right. I did not look to much into it, that's why I made the unecessary recursive loop. So I removed the `load_from_manifest` , `find_all_local_dependencies`, `build_with_deps` and only use the local packages retrieved from the main `pixi.lock`

Commit:
fix: remove recursive search and use packages from lock file

fix: fix typo

fix: fix typo

pavelzw commented on src/pack.rs:
```diff
@@ -179,8 +245,10 @@ pub async fn pack(options: PackOptions) -> Result<()> {
-    let injected_packages: Vec<(PathBuf, ArchiveType)> = options
-        .injected_packages
+    let mut all_injected_packages = loaded_injected_packages;
+    all_injected_packages.extend(options.injected_packages.iter().cloned());
```
How about we use `iter` and `chain`?

pavelzw commented on src/pack.rs:
```diff
@@ -119,14 +148,51 @@ pub async fn pack(options: PackOptions) -> Result<()> {
+                    output_folder.path(),
+                    options.platform.as_str(),
+                )
+                .await?;
```
Maybe out of scope for this PR but I'm wondering whether it makes sense to parallelize this.
@hofer-julian do you know whether there might be concurrency issues when invoking pixi-build for multiple packages at once?

pavelzw commented on src/pack.rs:
```diff
@@ -89,6 +90,31 @@ fn load_lockfile(manifest_path: &Path) -> Result<LockFile> {
+        .stdout(std::process::Stdio::null())
+        .stderr(std::process::Stdio::null())
+        .status()
+        .await?;
```
i think especially if we failed to build, it would be nice to see stdout/stderr (or stdout+stderr)

Also if we do `-v` I think it would be nice to show the output of pixi-build.

pavelzw commented on src/pack.rs:
```diff
@@ -209,6 +277,11 @@ pub async fn pack(options: PackOptions) -> Result<()> {
             .await
             .map_err(|e| anyhow!("could not copy file to channel directory: {}", e))?;
 
+        if let Some(parent) = path.parent()
```
I think this should always be `Some` so we can also use `path.parent().expect("Parent exists")`

pavelzw commented on src/pack.rs:
```diff
@@ -209,6 +277,11 @@ pub async fn pack(options: PackOptions) -> Result<()> {
+        if let Some(parent) = path.parent()
+            && parent == output_folder.path()
+        {
+            fs::remove_file(&path).await.ok();
```
I don't like this. How about building to a temporary directory instead and letting the `Drop` implementation remove it automatically?

pavelzw reviewed (commented):
Looks very solid already, thanks!
I have a couple of comments

pavelzw commented on src/pack.rs:
```diff
@@ -89,6 +90,31 @@ fn load_lockfile(manifest_path: &Path) -> Result<LockFile> {
+    let manifest_dir = manifest_path
+        .parent()
+        .ok_or_else(|| anyhow!("could not get parent directory of manifest_path"))?;
+    let status = TokioCommand::new("pixi")
```
Now that we invoke `pixi` here, I think we should add `pixi` as well in our `pixi.toml` for reproducibility's sake 😅 

When running `pixi run cargo test`, it should then use the pixi from the environment from `PATH` when using `pixi-build`.

Commit:
fix: fix after review

fix: fix typo for ci precommit

fix: fix typo for ci precommit

fix: fix typo for ci precommit

pavelzw committed:
refactor local-build structure

pavelzw committed:
explicitly drop

pavelzw committed:
reword

Commit:
fix: fix manifest_dir

fix: fix typo

dependabot[bot] committed:
ci: bump the gh-actions group with 5 updates (#247)

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

pavelzw committed:
build: Bump rattler, use uv from crates.io (#243)

Commit:
fix: fix tests to match pixi 0.6.1

regenerate lock file to use pixi 0.61.0

fix: pin the build backend issue

toprinse commented:
Hello @pavelzw.

So I looked into the issue of the regenerated hashes and it seems that recently the cmake build backend used in the tests has been updated and the hash generation might have been impacted. I'm not sure about how it impacted the tests and it's still unclear to me but for now I added the new generated hashes using the cmake build backend 0.3.5 (maybe it could be intersting to pin the build backend version in the test to avoid new strange behaviors).

Another idea could be to use match patterns instead of the exact version of the package containing the hashes. Instead of looking for `conda-meta/local-build-main-pkg-0.1.0-hb0f4dca_0.json` we could also use something like `starts_with("local-build-main-pkg-0.1.0-")` inside the conda-meta directory.

I also updated the pixi version to use 0.61.0 so the API of `pixi build` command has changed a bit in `pack.rs`

toprinse committed:
Merge branch 'main' into add_source_pkg_handling

Commit:
fix: fix cargo lock

pavelzw commented:
we now cleaned up and merged all other PRs and this one is next, sorry to keep you waiting for so long.

pixi lockfile v7 are coming soon-ish (https://github.com/prefix-dev/pixi/pull/5607, https://github.com/conda/rattler/pull/2026) and will lead to the `version` field not being fully required anymore.
we need to
- merge `main` into this branch and resolve conflicts
- see how the new lockfile format affects this PR.

Commit:
merge main into add_source_pkg_handling

Commit:
feat: add source package handling with cross-platform fixes

toprinse commented:
Hello @pavelzw . 
So I merged the main branch into this one, fixed the conflicts and update tests.
I made a few changes : 
- I removed the win64 test (see comments in `integrations_tests.rs` because they were failing because the generated path from pixi build is too long. There is currently no clean solutions to avoid reaching these windows limitations.
- Then, I'm not sure why but the `test_pixi_pack_source` sometimes fail, even I don't add major changes. I then get :
```bash
---- test_pixi_pack_source stdout ----
⏳ Downloading 25 packages...

thread 'test_pixi_pack_source' (13156) panicked at tests/integration_test.rs:858:5:
Err(could not download package: error decoding response body)
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

Maybe you have a better idea about what's happening ?

pavelzw commented:
> Then, I'm not sure why but the test_pixi_pack_source sometimes fail, even I don't add major changes

this looks like a networking issue and the package got dropped. maybe we need to add retries to the `Client` if we haven't done so already

pavelzw commented:
Lockfile v7 is now on main (of pixi). This means we should be able to test against it :)

pavelzw commented:
Added support for lockfile v7 in #306. hopefully this fixes our reproducibility issues

toprinse commented:
Hello @pavelzw. Sorry for the late answer. It's nice to have lockfiles v7 available !
Is there still things I can do for this PR ?

pavelzw commented:
I think we would need to merge main and then check again whether the tests would now be working reliably

Commit:
Merge remote-tracking branch 'upstream/main' into add_source_pkg_handling

Commit:
fix: fix typo

toprinse commented:
Hello @pavelzw. So, I merged `main` into this branch and fixed the breaking changes introduced by the `rattler_lock` refactor (specifically, CondaSourceData.package_record was removed — replaced by record() returning an Option<&PackageRecord>). The tests appear to be passing now. Still I'm not sure they are working reliably. Seems that the merge did not affect packages names and hashes in the test.

delsner committed:
Merge branch 'main' of github.com:Quantco/pixi-pack into add_source_pkg_handling

delsner committed:
Bump pixi

delsner committed:
Adjust implementation for most recent pixi publish

delsner committed:
Fix

delsner committed:
nit

pavelzw commented on src/pack.rs:
```diff
@@ -100,6 +102,77 @@ fn load_lockfile(manifest_path: &Path) -> Result<LockFile> {
+            .status()
+            .await?;
+
+        if !status.success() {
+            anyhow::bail!("Failed to publish package in {}", manifest_dir.display());
+        }
```
in a follow-up we could stream the logs like `pixi install`

pavelzw replied:
https://github.com/Quantco/pixi-pack/issues/330

pavelzw commented on src/pack.rs:
```diff
@@ -147,14 +220,78 @@ pub async fn pack(options: PackOptions) -> Result<()> {
+                };
+                let relative_path = match &source_data.location {
+                    UrlOrPath::Path(p) => PathBuf::from(p.to_path().as_str()),
+                    UrlOrPath::Url(u) => {
+                        anyhow::bail!("Unexpected URL for local package source: {}", u)
+                    }
```
this could be useful as well in a follow-up to support things like this:

```toml
[workspace]
preview = ["pixi-build"]
channels = ["conda-forge"]
platforms = ["linux-64", "linux-aarch64", "win-64", "osx-64", "osx-arm64"]

[dependencies]
agent-skill-conda-forge = { git = "https://github.com/pavelzw/skill-forge", rev = "main", subdirectory = "recipes/conda-forge" }
```

pavelzw replied:
https://github.com/Quantco/pixi-pack/issues/331

delsner committed:
fix?

delsner reviewed (approved):
Thanks for your work! Sorry that it took us a while to get this through.

delsner committed:
Disable test again

pavelzw committed:
Merge branch 'main' into add_source_pkg_handling

pavelzw committed:
bump
